test/test_jit_cuda_fuser.py

```diff
@@ -935,48 +935,39 @@ def test_binary_ops(self):
     @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING,
                      "Requires fusion optimization pass to be effective")
     def test_binary_bitwise(self):
-        def jit_or(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
-            return (x & y) | z
-
-        def jit_xor(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
-            return (x & y) ^ z
-
-        def jit_lshift(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
-            return (x & y) << z
-
-        def jit_rshift(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
-            return (x & y) >> z
-
-        for jit_func in [jit_or, jit_xor, jit_lshift, jit_rshift]:
-            x = torch.randn(4, 8, 32, 32, dtype=torch.float, device="cuda").mul(5).to(torch.long)
-            y = torch.randn(4, 8, 32, 32, dtype=torch.float, device="cuda").mul(5).to(torch.long)
-            z = torch.randn(4, 8, 32, 32, dtype=torch.float, device="cuda").mul(2).to(torch.long)
-
-            jitted = torch.jit.script(jit_func)
-            jit_o = jitted(x, y, z)
-            jit_o = jitted(x, y, z)
-            o = jit_func(x, y, z)
-            self.assertEqual(o, jit_o)
-            self.assertGraphContains(jitted.graph_for(x, y, z), FUSION_GUARD)
-
-        # We shouldn't need this redefinition of the function, but otherwise it won't recompile for a new type
-        def jit_or(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
-            return (x & y) | z
-
-        def jit_xor(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
-            return (x & y) ^ z
-
-        for jit_func in [jit_or, jit_xor]:
-            x = torch.rand(4, 2, dtype=torch.float, device="cuda").round().to(torch.bool)
-            y = torch.rand(4, 2, dtype=torch.float, device="cuda").round().to(torch.bool)
-            z = torch.rand(4, 2, dtype=torch.float, device="cuda").round().to(torch.bool)
-
-            jitted = torch.jit.script(jit_func)
-            jit_o = jitted(x, y, z)
-            jit_o = jitted(x, y, z)
-            o = jit_func(x, y, z)
-            self.assertEqual(o, jit_o)
-            self.assertGraphContains(jitted.graph_for(x, y, z), FUSION_GUARD)
+        dtypes = [torch.bool, torch.int32, torch.int64]
+
+        for dtype1, dtype2, dtype3 in itertools.product(dtypes, repeat=3):
+            def jit_and(x: torch.Tensor, y: torch.Tensor, z: torch.Tensor):
```

Comment:
I wish I could put it outside the dtype for loop, but unfortunately, it will fail if I do so, probably due to the recompile issue above?

Reply:
What recompile issue? If you are talking about caching for optimization engine, you can try this one https://github.com/csarofeen/pytorch/blob/devel/test/test_jit_cuda_fuser.py#L2632-L2633

Reply:
If it is only just about testing, then I think defining these functions inside the `for dtype...` loop is clean enough. So I won't update this PR. But if in some later PR I need to copy-paste the code to workaround the cache, then I will use this magic flag :)